### DISCUSSION TITLE
VC Service: First Milestone Considerations

### DISCUSSION THREAD

**Discussion by thereisnogabe** [TBD54566975/open-source-programs#15] [General]
## Goals

Support the following functional use case:

1. Everyone can have a DID-based identity
2. Sign and verify data with DID-based keys
3. Create and use JSON Schemas with Verifiable Credentials
4. Issue, hold, and verify VCs using DIDs and JSON Schemas
5. Apply for credential issuance via Manifest
6. Request and exchange credentials with Presentation Exchange

All this is able to be done via a command line and a simple REST/HTTP based service that uses in-memory or flat-file storage. There is no expectation of privacy or confidentiality of the information generated and exchanged, it is simply for demonstration purposes.

## Primitives

Standards-based primitives to reach the goals.

- DIDs
    - Using `did:key`
    - Create and generate `did:key` values of a few different types (Ed25519, secp256k1, ECDSA curves)
    - **Bonus:** consider `did:pkh` to connect existing blockchain accounts
- Credentials
    - Create and issue credentials using `did:key`s
    - Schema support
        - LD contexts with the JsonWebSignature2020 suite
        - JSON Schemas with JWS signing
- Schemas
    - Able to build and use JSON Schemas for Verifiable Credential data
- Signing & Verification
    - Support for the JsonWebSignature2020 type for Linked Data credentials
    - Support for JWS packaging for all else
- Presentation Exchange
    - Able to define simple (e.g. a single credential, from a single issuer, of a known schema) presentation requests
    - Able to fulfill those requests with a valid presentation response
- Credential Manifest
    - Able to describe simple manifests for credential issuance
    - Able to accept credentials (self signed are fine) to be issued a credential

## Non-Goals

- No authentication
- No service reliability guarantees (e.g. zero down time, multi instance, databases)
- No credential revocations
- No robust LD support
- No key rotation
- No assumption of wallet (plain text is fine)

What do you think?

**Comment by mistermoe** [comment]
@decentralgabe 

>3. Create and use JSON Schemas with Verifiable Credentials

Are you talking about JSON schemas for `credentialSubject` or are you thinking JSON schema to validate an entire Verifiable Credential as per spec? My guess is that you're going to say both 😄

**Reply by thereisnogabe** [reply]
Both!

1. Use JSON Schemas to validate _all_ our objects: credentials, dids, presentations, are complaint with the spec.
2. Further, use JSON Schemas specifically for the `credentialSubject` to enforce `type` (e.g. name schema, person schema)
3. We can apply this pattern to any type of object we send - like a DIDComm message of a certain shape

**Reply by mistermoe** [reply]
Do you think those `credentialSubject` specific JSON schemas will map 1:1 to the "forms" people will fill out to apply for a VC? My guess is probably not. Lmk if my question doesn't make sense and I'll try to clarify

**Reply by thereisnogabe** [reply]
Perhaps not 1:1 but the form will populate the `credentialSubject` field. It's the "meat" of the credential.

**Comment by ALRubinger** [comment]
This sounds awesome. I think it's really going to let new folks know what the vc-service is all about. I bet you'll file a bunch of new tasks/features as Issues in the process for others to pick up later.

In service of helping you find the tightest possible scope:

> command line and a simple REST/HTTP based service

`curl` or `httpie` instead of a CLI client, right?

> few different types (Ed25519, secp256k1, ECDSA curves)

One type if that saves you time and fills the E2E case

Non-Goals section is on point. Can't wait to see this come together.

**Reply by thereisnogabe** [reply]
> curl or httpie instead of a CLI client, right?

correct.

for key types - already done 😄

**Reply by ALRubinger** [reply]
*of course* it is. 😆 🔥

**Comment by maheshbalan** [comment]
WHat about the storage of credential schemas ?. Will that be in the ION network ?. Is the idea to support multiple DID methods ultimately to support interoperability?

**Reply by thereisnogabe** [reply]
schemas will be stored in DWNs for sure, and other places (perhaps IPFS) too.
Multiple DID methods - yes!

FYI @maheshbalan we've moved to https://forums.tbd.website/, if you have further discussion points please consider opening a thread there
